cmd/fs-v1-metadata.go

```diff
@@ -274,167 +259,109 @@ func newFSMetaV1() (fsMeta fsMetaV1) {
-	}
-
-	// If we couldn't read anything, The disk is unformatted.
-	if errorCause(err) == io.EOF {
-		err = errUnformattedDisk
-		format = newFSFormatV2()
-	} else {
-		// Validate loaded `format.json`.
-		err = checkFormatFS(format, fsFormatVersion)
-		if err != nil && err != errFSFormatOld {
-			return traceError(fmt.Errorf("Unable to validate 'format.json', %s", err))
-		}
-	}
-
-	// Disk is in old format migrate object metadata.
-	if err == errFSFormatOld {
-		if merr := migrateFSObject(fsPath, fsUUID); merr != nil {
-			return merr
-		}
-
-		// Initialize format v2.
-		format = newFSFormatV2()
-	}
+	fsFormatPath := pathJoin(fsPath, minioMetaBucket, formatConfigFile)
+
+	// Validate the `format.json` again after the initialization
+	// to further validate `format.json` entries to be expected.
+	//
+	// Once the filesystem has initialized hold the read lock for
+	// the life time of the server. This is done to ensure that under
+	// shared backend mode for FS, remote servers do not migrate
+	// or cause changes on backend format. There is a slight window
+	// for this to occur.
+
+	// This loop validates format.json by holding a read lock and
+	// proceeds to holding a TryWriteLock by validating if the
+	// disk is unformatted. If for some reason TryWriteLock fails
+	// and the error returned is lock.ErrAlreadyLocked i.e some other
+	// process is holding a lock we make sure that we try again
+	// in the loop by holding a read lock.
+	var rlk *lock.RLockedFile
+	for {
+		rlk, err = checkLockedValidFormatFS(fsPath)
+		switch {
+		case err == nil:
+			// Holding a read lock ensures that any write lock operation
+			// is blocked if attempted in-turn avoiding corruption on
+			// the backend disk.
+			_ = rlk // Hold the lock on `format.json` until server dies.
+			return nil
+		case errorCause(err) == errUnformattedDisk:
+			// Attempt a write lock on formatConfigFile `format.json`
+			// file stored in minioMetaBucket(.minio.sys) directory.
+			lk, lerr := lock.TryLockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
+			if lerr != nil {
+				// Existing write locks detected.
+				if lerr == lock.ErrAlreadyLocked {
+					// Lock already present, sleep and attempt again.
+					time.Sleep(100 * time.Millisecond)
+					continue
+				}
+				// Unexpected error, return.
+				return traceError(lerr)
+			}
 
-	// Rewrite or write format.json depending on if disk
-	// unformatted and if format is old.
-	if err == errUnformattedDisk || err == errFSFormatOld {
-		if _, err = format.WriteTo(lk); err != nil {
-			return traceError(fmt.Errorf("Unable to initialize 'format.json', %s", err))
+			// Load format to validate again, checks if we are unformatted
+			// writes the new format.json
+			var format = &formatConfigV1{}
+			lerr = format.LoadFormat(lk)
+			switch {
+			case lerr == nil:
+				// Formatting completed by another process, continue
+				// to attempt a read-lock and validate `format.json`.
+				lk.Close()
+				continue
```

Comment:
lk.Close() can be closed after switch statement - this simplifies a couple of cases here.